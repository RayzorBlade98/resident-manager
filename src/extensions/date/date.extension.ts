declare interface Date {
  /**
   * Converts the date to a new date in UTC timezone
   */
  toUTC(): Date;

  /**
   * Converts the date to 'DD.MM.YYYY' format string
   */
  toPreferredString(): string;
}

function toUTC(this: Date): Date {
  return new Date(
    Date.UTC(this.getFullYear(), this.getMonth(), this.getDate(), 0, 0, 0, 0),
  );
}
Date.prototype.toUTC = toUTC;

function toPreferredString(this: Date): string {
  const day: number = this.getDate();
  const month: number = this.getMonth() + 1;
  const year = String(this.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
Date.prototype.toPreferredString = toPreferredString;
