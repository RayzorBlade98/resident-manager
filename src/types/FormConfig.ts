import { GroupedFormProps } from '_/components/form/GroupedForm/GroupedForm';
import { FormValidationArguments } from '_/hooks/useFormValidation/useFormValidation';

/**
 * Object containing the config of a grouped form
 */
export type FormConfig<TInput extends object, TGroups extends string> = {
  formValidationConfig: Omit<
  FormValidationArguments<TInput>,
  'onSubmitSuccess'
  >;
  formGroupConfig: Omit<
  GroupedFormProps<TInput, TGroups>,
  'children' | 'formErrors'
  >;
};
