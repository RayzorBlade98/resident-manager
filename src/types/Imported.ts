import Prettify from './Prettify';

/**
 * Type that replaces all date types with the string type
 * representing the result of the import of an object from type `T`
 */
type Imported<T extends object> = Prettify<{
  [key in keyof T]: T[key] extends Date | undefined
    ? string
    : T[key] extends object
      ? Imported<T[key]>
      : T[key];
}>;

export default Imported;
