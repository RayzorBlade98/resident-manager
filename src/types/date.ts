/**
 * Enum containing all months as `string`
 */
export enum Month {
  January = 'Januar',
  Febuary = 'Februar',
  March = 'MÃ¤rz',
  April = 'April',
  May = 'Mai',
  June = 'Juni',
  July = 'Juli',
  August = 'August',
  September = 'September',
  October = 'Oktober',
  November = 'November',
  December = 'Dezember',
}

/**
 * Type representing a year
 */
export type Year = number;

/**
 * Information about a specific month and year
 */
export interface MonthYear {
  month: Month;
  year: Year;
}

/**
 * Type of the string representation of a `MonthYear`
 *
 * Format: "[month] [year]"
 */
export type MonthYearString = `${Month} ${Year}`;

/**
 * Type of the string representation of a date including day, month and year
 *
 * Format: `dd.mm.yyyy`
 */
export type DateString = `${string}.${string}.${string}`;

/**
 * Type of the date string representation returned by the date input field
 *
 * Format: `yyyy-mm-dd`
 */
export type DateInputString = `${string}-${string}-${string}`;

/**
 * Class that provides utility functions for `MonthYear` objects
 */
export abstract class MonthYearUtils {
  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param monthYear `MonthYear` object to which the months should be added
   * @param addedMonths Number of months that should be added (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static addMonths(monthYear: MonthYear, addedMonths = 1): MonthYear {
    if (addedMonths < 0) {
      // Add negative amount of months == subtract months
      return MonthYearUtils.subtractMonths(monthYear, addedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.December) {
      // Year change
      newMonthYear.month = Month.January;
      newMonthYear.year += 1;
    } else {
      // Add one month to the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex + 1];
    }

    // Call `addMonths` recursive until all months got addded
    if (addedMonths !== 1) {
      return MonthYearUtils.addMonths(newMonthYear, addedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Checks if two `MonthYear` objects are equal
   * @param a first `MonthYear` object to compare
   * @param b second `MonthYear` object to compare
   * @returns `true` if both the month and year are the same, else `false`
   */
  public static areEqual(a: MonthYear, b: MonthYear): boolean {
    return a.month === b.month && a.year === b.year;
  }

  /**
   * Compares two `MonthYear` objects
   * @param a First `MonthYear` object to compare
   * @param b Second `MonthYear` object to compare
   * @returns
   *
   * - 0 if both object have the same month and year
   * - <0 if the first object contains a date that's earlier than the second object
   * - \>0 if the first object contains a date that's later than the second object
   */
  public static compare(a: MonthYear, b: MonthYear): number {
    if (a.year !== b.year) return a.year - b.year;

    const months = Object.values(Month);
    const aMonthIndex = months.findIndex((m: Month) => m === a.month);
    const bMonthIndex = months.findIndex((m: Month) => m === b.month);
    return aMonthIndex - bMonthIndex;
  }

  /**
   * Returns a `MonthYear` object that contains information of the current month and year
   * @returns `MonthYear` object for the current date
   */
  public static getCurrentMonthYear(): MonthYear {
    const date: Date = new Date();
    const monthIndex: number = date.getUTCMonth();
    const month: Month = Object.values(Month)[monthIndex] as Month;
    const year: Year = date.getUTCFullYear();

    return {
      month,
      year,
    };
  }

  /**
   * Parses a string representation of a `MonthYear` object to its respective object
   * @param monthYear string representation of a `MonthYear` object that should be parsed
   * @returns `MonthYear` object that contains the information of the specified `string`
   */
  public static parseString(monthYear: MonthYearString): MonthYear {
    const splits = monthYear
      .split(/(\s+)/)
      .filter((s: string) => s.trim().length > 0);
    return {
      month: splits[0] as Month,
      year: Number(splits[1]),
    };
  }

  /**
   * Subtracts the specified number of months from the given `MonthYear` object
   * @param monthYear `MonthYear` object from which the months should be subtraced
   * @param subtractedMonths Number of months that should be subtracted (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static subtractMonths(
    monthYear: MonthYear,
    subtractedMonths = 1,
  ): MonthYear {
    if (subtractedMonths < 0) {
      // Subtract negative amount of months == add months
      return MonthYearUtils.addMonths(monthYear, subtractedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.January) {
      // Year change
      newMonthYear.month = Month.December;
      newMonthYear.year -= 1;
    } else {
      // Subtract one month from the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex - 1];
    }

    // Call `subtractMonths` recursive until all months got subtracted
    if (subtractedMonths !== 1) {
      return MonthYearUtils.subtractMonths(newMonthYear, subtractedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Converts a `MonthYear` to its string representation
   * @param monthYear `MonthYear` that should be converted
   * @returns string representation of the specified `MonthYear` object with the format "[month] [year]"
   */
  public static toString(monthYear: MonthYear): MonthYearString {
    return `${monthYear.month} ${monthYear.year}`;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    const timespan = [{ ...start }];

    // Whether to add a month each loop or subtract one
    const direction = Math.sign(MonthYearUtils.compare(end, start));

    if (MonthYearUtils.areEqual(start, end)) return timespan;

    for (
      let next = MonthYearUtils.addMonths(start, direction);
      !MonthYearUtils.areEqual(next, end);
      next = MonthYearUtils.addMonths(next, direction)
    ) {
      timespan.push(next);
    }
    timespan.push({ ...end });
    return timespan;
  }
}

/**
 * Class that provides utility functions for dates containing day, month and year.
 */
export abstract class DateUtils {
  /**
   * Converts a date string from `yyyy-mm-dd` to `dd.mm.yyyy` format
   * @param date Date in `yyyy-mm-dd` format
   * @returns Converted date in `dd.mm.yyyy` format
   */
  public static convertDateInputString(date: DateInputString): DateString {
    const splits = date.split('-');
    return `${splits[2]}.${splits[1]}.${splits[0]}`;
  }

  /**
   * Converts a date string from `dd.mm.yyyy` to `yyyy-mm-dd` format
   * @param date Date in `dd.mm.yyyy` format
   * @returns Converted date in `yyyy-mm-dd` format
   */
  public static convertToDateInputString(date: DateString): DateInputString {
    const splits = date.split('.');
    return `${splits[2]}-${splits[1]}-${splits[0]}`;
  }

  /**
   * Returns a date string containing the current date
   * @returns Date string containing the current date in `dd.mm.yyyy` format
   */
  public static getCurrentDate(): DateString {
    const date: Date = new Date();
    const day: number = date.getUTCDate();
    const month: number = date.getUTCMonth() + 1;
    const year = String(date.getUTCFullYear());

    const dayPrefix: string = day < 10 ? '0' : '';
    const monthPrefix: string = month < 10 ? '0' : '';

    return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
  }
}
