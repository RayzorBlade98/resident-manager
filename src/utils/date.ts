// Remove after adding another export:
/* eslint-disable import/prefer-default-export */

/**
 * Converts the given date to its string representaion (format: DD.MM.YYYY)
 * @param date date that should be converted
 * @returns string representation of the date
 */
export function dateToString(date: Date): string {
  const day: number = date.getDate();
  const month: number = date.getMonth() + 1;
  const year = String(date.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
