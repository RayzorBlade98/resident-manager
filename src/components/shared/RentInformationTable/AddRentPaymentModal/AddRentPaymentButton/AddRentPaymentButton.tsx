import React from 'react';
import { useRecoilValue, useResetRecoilState } from 'recoil';
import FormSubmitButton from '_/components/form/FormSubmitButton/FormSubmitButton';
import addRentPaymentState, {
  RentPaymentInput,
  addRentPaymentFormValidationSelector,
} from '_/components/shared/RentInformationTable/states/add_rent_payment_state';
import MonthYear from '_/extensions/date/month_year.extension';
import { Resident } from '_/models/resident/resident';
import ResidentStateManager from '_/states/resident/resident.state.manager';
import '_/extensions/date/date.extension';

/**
 * Button that submits the input payment infos if they are valid
 */
function AddRentPaymentButton(): JSX.Element {
  const rentPaymentState = useRecoilValue(addRentPaymentState);
  const resetRentPaymentState = useResetRecoilState(addRentPaymentState);

  const onSuccess = (): void => {
    ResidentStateManager.updateRentInformation(
      (rentPaymentState.selectedResident as Resident).id,
      rentPaymentState.selectedRentMonth as MonthYear,
      {
        paymentAmount: rentPaymentState.formValidation.formInput.paymentAmount,
        paymentDate:
          rentPaymentState.formValidation.formInput.paymentDate?.toUTC(),
      },
    );
    resetRentPaymentState();
  };

  return (
    <FormSubmitButton<RentPaymentInput>
      buttonText="HinzufÃ¼gen"
      formState={addRentPaymentFormValidationSelector}
      onSuccess={onSuccess}
    />
  );
}

export default AddRentPaymentButton;
