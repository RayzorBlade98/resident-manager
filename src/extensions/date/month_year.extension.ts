import './date.extension';

const MONTH_NAMES = [
  'Januar',
  'Februar',
  'MÃ¤rz',
  'April',
  'Mai',
  'Juni',
  'Juli',
  'August',
  'September',
  'Oktober',
  'November',
  'Dezember',
];

/**
 * Date object that represents a month and a year
 */
class MonthYear extends Date {
  /**
   * Creates a new `MonthYear` object
   * @param month month of the new object. If undefined the current month will be used.
   * @param year year of the new object. If undefined the current year will be used.
   */
  constructor(month?: number, year?: number) {
    const now = new Date();
    const utc = new Date(
      year !== undefined ? year : now.getFullYear(),
      month !== undefined ? month : now.getMonth(),
      1,
    ).toUTC();
    super(utc.getTime());
  }

  /**
   * Converts a `Date` object to a `MonthYear`
   * @param date Date that should be converted
   * @returns `MonthYear` with the same month and year as `date`
   */
  public static fromDate(date: Date): MonthYear {
    return new MonthYear(date.getMonth(), date.getFullYear());
  }

  /**
   * Converts a date string to a `MonthYear`
   * @param date Date string that should be converted
   * @returns `MonthYear` with the same month and year as `date`
   */
  public static fromString(date: string): MonthYear {
    return MonthYear.fromDate(new Date(date));
  }

  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param toAdd Number of months that should be added (can be negative)
   * @returns this month year
   */
  public addMonths(toAdd = 1): MonthYear {
    this.setMonth(this.getMonth() + toAdd);
    return this;
  }

  /**
   * Creates a copy this object.
   * @returns the copy with the same month and year
   */
  public clone(): MonthYear {
    return new MonthYear(this.getMonth(), this.getFullYear());
  }

  /**
   * Checks whether this object has the same month and year as anoter `MonthYear`
   * @param other true if both the month and year are the same, else false
   */
  public equals(other: MonthYear): boolean {
    return this >= other && this <= other;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    // Whether to add a month each loop or subtract one
    const direction = end > start ? 1 : -1;

    const timespan: MonthYear[] = [start.clone()];
    let next = start;
    while (!end.equals(next)) {
      next = next.clone();
      next.addMonths(direction);
      timespan.push(next);
    }

    return timespan;
  }

  /**
   * Generates a string representation in the format `<month name> <year>`
   * @returns string representation of this object
   */
  public toString(): string {
    return `${MONTH_NAMES[this.getMonth()]} ${this.getFullYear()}`;
  }
}

export default MonthYear;
