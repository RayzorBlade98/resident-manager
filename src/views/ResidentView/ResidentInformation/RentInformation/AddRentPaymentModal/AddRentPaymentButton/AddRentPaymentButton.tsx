import { Button } from '@mui/material';
import React from 'react';
import { useRecoilState, useRecoilValue, useResetRecoilState } from 'recoil';
import addRentPaymentState, {
  paymentValidator,
} from '../../states/add_rent_payment_state';
import { ResidentStateManager } from '_/states/saveStates/resident_state';
import { MonthYear } from '_/types/date';
import { Resident } from '_/types/resident';
import { CurrencyInCents } from '_/utils/currency';
import { residentViewSelectedResidentState } from '_/views/ResidentView/states/resident_view_state';

/**
 * Button that submits the input payment infos if they are valid
 */
function AddRentPaymentButton(): JSX.Element {
  // eslint-disable-next-line max-len
  const [rentPaymentState, setRentPaymentState] = useRecoilState(addRentPaymentState);
  const resetRentPaymentState = useResetRecoilState(addRentPaymentState);

  const selectedResident = useRecoilValue(
    residentViewSelectedResidentState,
  ) as Resident;

  function onSave(): void {
    const errors = paymentValidator.validate(rentPaymentState.formInput);

    if (Object.keys(errors).length > 0) {
      setRentPaymentState((state) => ({ ...state, formErrors: errors }));
      return;
    }

    ResidentStateManager.updateRentInformation(
      selectedResident.id,
      rentPaymentState.selectedRentMonth as MonthYear,
      {
        paymentAmount: rentPaymentState.formInput
          .paymentAmount as CurrencyInCents,
        paymentDate: rentPaymentState.formInput.paymentDate,
      },
    );
    resetRentPaymentState();
  }

  return (
    <Button variant="contained" onClick={() => onSave()}>
      Hinzuf√ºgen
    </Button>
  );
}

export default AddRentPaymentButton;
