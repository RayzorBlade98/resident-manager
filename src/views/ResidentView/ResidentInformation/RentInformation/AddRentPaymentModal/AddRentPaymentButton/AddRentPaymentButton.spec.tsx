/* eslint-disable max-len */

import {
  RenderResult, act, fireEvent, render,
} from '@testing-library/react';
import React from 'react';
import * as RecoilModule from 'recoil';
import { getRecoil, setRecoil } from 'recoil-nexus';
import addRentPaymentState, {
  addRentPaymentFormValidationSelector,
} from '../../states/add_rent_payment_state';
import AddRentPaymentButton from './AddRentPaymentButton';
import residentState, {
  ResidentStateManager,
} from '_/states/saveStates/resident_state';
import { Month, MonthYear } from '_/types/date';
import residentViewState from '_/views/ResidentView/states/resident_view_state';
import RecoilTestWrapper from '_tests/__test_utils__/RecoillTestWrapper';
import ResidentBuilder from '_tests/__test_utils__/builders/resident_builder';
import '_/extensions/date/date.extension';

describe('AddRentPaymentButton', () => {
  const selectedResident = new ResidentBuilder().build();
  const selectedRentMonth: MonthYear = {
    month: Month.May,
    year: 2023,
  };
  const validInputValues = {
    paymentAmount: 100,
    paymentDate: new Date(26, 4, 2023).toUTC(),
  };

  let renderResult: RenderResult;
  let updateRentInformationSpy: jest.SpyInstance;
  let resetRentPaymentStateSpy: jest.Mock;

  function validInput(): void {
    act(() => {
      setRecoil(addRentPaymentState, (state) => ({
        ...state,
        formValidation: {
          ...state.formValidation,
          formInput: { ...validInputValues },
        },
      }));
    });
  }

  function invalidInput(): void {
    act(() => {
      setRecoil(addRentPaymentState, (state) => ({
        ...state,
        formValidation: {
          ...state.formValidation,
          formInput: {
            ...state.formValidation.formInput,
            paymentAmount: undefined,
          },
        },
      }));
    });
  }

  function pressButton(): void {
    const button = renderResult.getByRole('button');
    fireEvent.click(button);
  }

  beforeEach(() => {
    updateRentInformationSpy = jest
      .spyOn(ResidentStateManager, 'updateRentInformation')
      .mockReturnValue(undefined);

    resetRentPaymentStateSpy = jest.fn();
    jest
      .spyOn(RecoilModule, 'useResetRecoilState')
      .mockReturnValue(resetRentPaymentStateSpy);

    renderResult = render(
      <RecoilTestWrapper>
        <AddRentPaymentButton />
      </RecoilTestWrapper>,
    );

    act(() => {
      setRecoil(residentState, [selectedResident]);
      setRecoil(residentViewState, (state) => ({
        ...state,
        selectedResident: selectedResident.id,
      }));
      setRecoil(addRentPaymentState, (state) => ({
        ...state,
        selectedRentMonth,
      }));
    });
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  test('should update rent information for valid inputs', () => {
    // Arrange
    validInput();

    // Act
    pressButton();

    // Assert
    expect(updateRentInformationSpy).toHaveBeenCalledTimes(1);
    expect(updateRentInformationSpy).toHaveBeenCalledWith(
      selectedResident.id,
      selectedRentMonth,
      validInputValues,
    );
  });

  test('should reset rent payment state for valid inputs', () => {
    // Arrange
    validInput();

    // Act
    pressButton();

    // Assert
    expect(resetRentPaymentStateSpy).toHaveBeenCalledTimes(1);
  });

  test('should not update rent information for invalid inputs', () => {
    // Arrange
    invalidInput();

    // Act
    pressButton();

    // Assert
    expect(updateRentInformationSpy).toHaveBeenCalledTimes(0);
  });

  test('should write error message to state for invalid inputs', () => {
    // Arrange
    invalidInput();

    // Act
    pressButton();

    // Assert
    const errorMessage = getRecoil(addRentPaymentFormValidationSelector)
      .formErrors.paymentAmount;
    expect(errorMessage).toBeDefined();
  });
});
