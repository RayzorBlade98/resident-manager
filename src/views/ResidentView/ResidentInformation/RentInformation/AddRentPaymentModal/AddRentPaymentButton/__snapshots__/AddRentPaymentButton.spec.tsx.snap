// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRentPaymentButton should match snapshot 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Hinzuf√ºgen
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
